File: src/Astbox.Core/CborDet.cs
// Copyright 2026 Astenyx-Git
// SPDX-License-Identifier: MPL-2.0 OR AGPL-3.0-only
//! Deterministic (canonical) CBOR encode/decode for ASTBOX v1 metadata.
//! Port of astbox/cbor_det.py (ASTBOX-v1.0-03 §13-19):
//! RFC 8949 deterministic encoding restricted to unsigned ints, byte strings,
//! UTF-8 text, arrays and maps; floats/tags/indefinite lengths/non-minimal
//! integers/duplicate or out-of-canonical-order map keys are rejected.

using System.Buffers.Binary;
using System.Text;

namespace Astbox;

/// <summary>Decoded CBOR value model (UInt | Bytes | Text | Array | Map).</summary>
public sealed class CborValue : IEquatable<CborValue>
{
    public enum Kinds : byte { UInt, Bytes, Text, Array, Map }

    public Kinds Kind { get; }
    public ulong AsUInt { get; }
    public byte[] AsBytes { get; }
    public string AsText { get; }
    public IReadOnlyList<CborValue> Items { get; }
    /// <summary>Map entries in decoded (canonical) order.</summary>
    public IReadOnlyList<KeyValuePair<CborValue, CborValue>> Entries { get; }

    private CborValue(Kinds kind, ulong u = 0, byte[]? bytes = null,
        string? text = null, IReadOnlyList<CborValue>? items = null,
        IReadOnlyList<KeyValuePair<CborValue, CborValue>>? entries = null)
    {
        Kind = kind;
        AsUInt = u;
        AsBytes = bytes ?? Array.Empty<byte>();
        AsText = text ?? string.Empty;
        Items = items ?? Array.Empty<CborValue>();
        Entries = entries ?? Array.Empty<KeyValuePair<CborValue, CborValue>>();
    }

    // ---- builders -------------------------------------------------------
    public static CborValue UInt(ulong v) => new(Kinds.UInt, u: v);
    public static CborValue Bytes(byte[] b) => new(Kinds.Bytes, bytes: b);
    public static CborValue Text(string s) => new(Kinds.Text, text: s);
    public static CborValue Arr(IReadOnlyList<CborValue> items)
        => new(Kinds.Array, items: items);
    public static CborValue Arr(params CborValue[] items)
        => new(Kinds.Array, items: items);
    public static CborValue Map(
        IReadOnlyList<KeyValuePair<CborValue, CborValue>> entries)
        => new(Kinds.Map, entries: entries);
    public static CborValue Map(params (ulong key, CborValue val)[] pairs)
        => new(Kinds.Map,
            entries: pairs.Select(p =>
                new KeyValuePair<CborValue, CborValue>(UInt(p.key), p.val))
                .ToList());

    public bool IsUInt => Kind == Kinds.UInt;
    public bool IsBytes => Kind == Kinds.Bytes;
    public bool IsText => Kind == Kinds.Text;
    public bool IsArray => Kind == Kinds.Array;
    public bool IsMap => Kind == Kinds.Map;

    public bool Equals(CborValue? other)
    {
        if (other is null || other.Kind != Kind) return false;
        return Kind switch
        {
            Kinds.UInt => AsUInt == other.AsUInt,
            Kinds.Bytes => AsBytes.AsSpan().SequenceEqual(other.AsBytes),
            Kinds.Text => string.Equals(AsText, other.AsText, StringComparison.Ordinal),
            Kinds.Array => Items.Count == other.Items.Count &&
                Items.Zip(other.Items, (a, b) => a.Equals(b)).All(x => x),
            Kinds.Map => Entries.Count == other.Entries.Count &&
                Entries.Zip(other.Entries, (a, b) =>
                    a.Key.Equals(b.Key) && a.Value.Equals(b.Value)).All(x => x),
            _ => false,
        };
    }

    public override bool Equals(object? obj) => Equals(obj as CborValue);

    public override int GetHashCode()
    {
        var h = new HashCode();
        h.Add(Kind);
        switch (Kind)
        {
            case Kinds.UInt: h.Add(AsUInt); break;
            case Kinds.Bytes:
                foreach (var b in AsBytes) h.Add(b);
                break;
            case Kinds.Text: h.Add(AsText, StringComparer.Ordinal); break;
            case Kinds.Array:
                foreach (var i in Items) h.Add(i); break;
            case Kinds.Map:
                foreach (var kv in Entries) { h.Add(kv.Key); h.Add(kv.Value); }
                break;
        }
        return h.ToHashCode();
    }

    public override string ToString()
        => Kind switch
        {
            Kinds.UInt => $"uint({AsUInt})",
            Kinds.Bytes => $"bytes[{AsBytes.Length}]",
            Kinds.Text => $"text({AsText.Length})",
            Kinds.Array => $"array[{Items.Count}]",
            Kinds.Map => $"map[{Entries.Count}]",
            _ => Kind.ToString(),
        };
}

public static class CborDet
{
    private const int MaxDepth = 64;
    private static readonly UTF8Encoding StrictUtf8 = new(false, true);

    // ------------------------------------------------------------------
    // Strict decoding
    // ------------------------------------------------------------------

    /// <summary>Strictly decode canonical CBOR; trailing bytes forbidden.</summary>
    public static CborValue Loads(ReadOnlySpan<byte> data)
    {
        var r = new Reader(data.ToArray());
        var obj = DecodeItem(r, 0);
        if (r.Remaining != 0)
            throw new AstboxError(E.InvalidCbor, "trailing bytes after CBOR item");
        return obj;
    }

    private sealed class Reader
    {
        public Reader(byte[] data) => Data = data;
        public byte[] Data { get; }
        public int Pos { get; set; }
        public int End => Data.Length;
        public int Remaining => End - Pos;

        public byte[] Take(int n)
        {
            if (Pos + n > End)
                throw new AstboxError(E.InvalidCbor, "truncated CBOR item");
            var outp = new byte[n];
            Buffer.BlockCopy(Data, Pos, outp, 0, n);
            Pos += n;
            return outp;
        }

        public ReadOnlySpan<byte> TakeSpan(int n)
        {
            if (Pos + n > End)
                throw new AstboxError(E.InvalidCbor, "truncated CBOR item");
            var s = Data.AsSpan(Pos, n);
            Pos += n;
            return s;
        }
    }

    // (major, additionalInfo, length)
    private static (int Major, int Ai, ulong Length) ReadHead(Reader r)
    {
        if (r.Remaining < 1)
            throw new AstboxError(E.InvalidCbor, "truncated CBOR item");
        int b0 = r.Data[r.Pos];
        r.Pos++;
        int major = b0 >> 5;
        int ai = b0 & 0x1F;
        switch (ai)
        {
            case < 24: return (major, ai, (ulong)ai);
            case 24: return (major, ai, r.Take(1)[0]);
            case 25: return (major, ai, BinaryPrimitives.ReadUInt16BigEndian(r.TakeSpan(2)));
            case 26: return (major, ai, BinaryPrimitives.ReadUInt32BigEndian(r.TakeSpan(4)));
            case 27: return (major, ai, BinaryPrimitives.ReadUInt64BigEndian(r.TakeSpan(8)));
            default:
                throw new AstboxError(E.InvalidCbor, "indefinite-length item forbidden");
        }
    }

    private static void CheckMinimal(int ai, ulong length)
    {
        // Reject non-minimal integer encodings (RFC 8949 4.2.1).
        if (ai == 24 && length < 24)
            throw new AstboxError(E.NonCanonicalCbor, "non-minimal uint encoding");
        if (ai == 25 && length < 0x100)
            throw new AstboxError(E.NonCanonicalCbor, "non-minimal uint encoding");
        if (ai == 26 && length < 0x10000)
            throw new AstboxError(E.NonCanonicalCbor, "non-minimal uint encoding");
        if (ai == 27 && length < 0x1_0000_0000)
            throw new AstboxError(E.NonCanonicalCbor, "non-minimal uint encoding");
    }

    private static CborValue DecodeItem(Reader r, int depth)
    {
        if (depth > MaxDepth)
            throw new AstboxError(E.InvalidCbor, "CBOR nesting too deep");
        if (r.Remaining < 1)
            throw new AstboxError(E.InvalidCbor, "truncated CBOR item");
        int major;
        int ai;
        ulong length;
        {
            int b0 = r.Data[r.Pos];
            (major, ai, length) = ReadHead(r);
            if (major != 0)
            {
                // rewind not needed; only uint uses minimal check below
            }
            else
            {
                CheckMinimal(ai, length);
                return CborValue.UInt(length);
            }
        }

        switch (major)
        {
            case 1: // negative integer: not permitted by ASTBOX metadata
                throw new AstboxError(E.InvalidCbor, "negative CBOR integer forbidden");

            case 2: // byte string
            {
                if (length > int.MaxValue)
                    throw new AstboxError(E.InvalidCbor, "byte string too large");
                return CborValue.Bytes(r.Take((int)length));
            }

            case 3: // text string
            {
                if (length > int.MaxValue)
                    throw new AstboxError(E.InvalidCbor, "text string too large");
                var raw = r.Take((int)length);
                string text;
                try { text = StrictUtf8.GetString(raw); }
                catch (DecoderFallbackException)
                {
                    throw new AstboxError(E.InvalidCbor, "text string is not UTF-8");
                }
                return CborValue.Text(text);
            }

            case 4: // array
            {
                if (length > int.MaxValue)
                    throw new AstboxError(E.InvalidCbor, "array too large");
                int n = (int)length;
                var arr = new List<CborValue>(n);
                for (int i = 0; i < n; i++)
                    arr.Add(DecodeItem(r, depth + 1));
                return CborValue.Arr(arr);
            }

            case 5: // map
            {
                if (length > int.MaxValue)
                    throw new AstboxError(E.InvalidCbor, "map too large");
                int n = (int)length;
                (int Len, byte[] Bytes)? prevEncoded = null;
                var result = new List<KeyValuePair<CborValue, CborValue>>(n);
                var seenKeys = new List<CborValue>(n);
                for (int i = 0; i < n; i++)
                {
                    int keyStart = r.Pos;
                    var key = DecodeItem(r, depth + 1);
                    int keyEnd = r.Pos;
                    var value = DecodeItem(r, depth + 1);
                    if (seenKeys.Any(k => k.Equals(key)))
                        throw new AstboxError(E.DuplicateCborKey,
                            $"duplicate map key {key}");
                    seenKeys.Add(key);
                    result.Add(new KeyValuePair<CborValue, CborValue>(key, value));
                    // canonical key order: keys sorted by (encoded length, bytes)
                    int keyLen = keyEnd - keyStart;
                    var keyBytes = r.Data[keyStart..keyEnd];
                    if (prevEncoded is { } prev &&
                        (keyLen < prev.Len ||
                         (keyLen == prev.Len &&
                          keyBytes.AsSpan().SequenceCompareTo(prev.Bytes) <= 0)))
                    {
                        throw new AstboxError(E.NonCanonicalCbor,
                            "map keys not in canonical order");
                    }
                    prevEncoded = (keyLen, keyBytes.ToArray());
                }
                return CborValue.Map(result);
            }

            // majors 6 (tags) and 7 (floats/specials) are forbidden
            default:
                throw new AstboxError(E.InvalidCbor,
                    $"CBOR major type {major} forbidden in ASTBOX metadata");
        }
    }

    // ------------------------------------------------------------------
    // Canonical encoding (subset used by ASTBOX metadata)
    // ------------------------------------------------------------------

    private static byte[] EncodeUint(ulong value)
    {
        if (value < 24) return [(byte)value];
        if (value < 0x100) return [0x18, (byte)value];
        if (value < 0x1_0000)
        {
            var b = new byte[3];
            b[0] = 0x19;
            BinaryPrimitives.WriteUInt16BigEndian(b.AsSpan(1), (ushort)value);
            return b;
        }
        if (value < 0x1_0000_0000)
        {
            var b = new byte[5];
            b[0] = 0x1A;
            BinaryPrimitives.WriteUInt32BigEndian(b.AsSpan(1), (uint)value);
            return b;
        }
        {
            var b = new byte[9];
            b[0] = 0x1B;
            BinaryPrimitives.WriteUInt64BigEndian(b.AsSpan(1), value);
            return b;
        }
    }

    private static byte[] EncodeHead(int major, ulong length)
    {
        if (length < 24) return [(byte)((major << 5) | (int)length)];
        if (length < 0x100) return [(byte)((major << 5) | 24), (byte)length];
        if (length < 0x1_0000)
        {
            var b = new byte[3];
            b[0] = (byte)((major << 5) | 25);
            BinaryPrimitives.WriteUInt16BigEndian(b.AsSpan(1), (ushort)length);
            return b;
        }
        if (length < 0x1_0000_0000)
        {
            var b = new byte[5];
            b[0] = (byte)((major << 5) | 26);
            BinaryPrimitives.WriteUInt32BigEndian(b.AsSpan(1), (uint)length);
            return b;
        }
        {
            var b = new byte[9];
            b[0] = (byte)((major << 5) | 27);
            BinaryPrimitives.WriteUInt64BigEndian(b.AsSpan(1), length);
            return b;
        }
    }

    private static void AppendItem(List<byte> acc, CborValue obj)
    {
        switch (obj.Kind)
        {
            case CborValue.Kinds.UInt:
                acc.AddRange(EncodeUint(obj.AsUInt));
                break;

            case CborValue.Kinds.Bytes:
                acc.AddRange(EncodeHead(2, (ulong)obj.AsBytes.Length));
                acc.AddRange(obj.AsBytes);
                break;

            case CborValue.Kinds.Text:
            {
                // metadata strings are NFC-normalized
                var norm = obj.AsText.Normalize(NormalizationForm.FormC);
                var data = StrictUtf8.GetBytes(norm);
                acc.AddRange(EncodeHead(3, (ulong)data.Length));
                acc.AddRange(data);
                break;
            }

            case CborValue.Kinds.Array:
            {
                acc.AddRange(EncodeHead(4, (ulong)obj.Items.Count));
                foreach (var item in obj.Items)
                    AppendItem(acc, item);
                break;
            }

            case CborValue.Kinds.Map:
            {
                var ordered = obj.Entries
                    .Select(kv => (
                        Key: kv.Key,
                        Value: kv.Value,
                        EncKey: EncodeKey(kv.Key)))
                    .OrderBy(t => t.EncKey, ByteArrayOrderComparer.Instance)
                    .ToList();
                acc.AddRange(EncodeHead(5, (ulong)ordered.Count));
                foreach (var t in ordered)
                {
                    acc.AddRange(t.EncKey);
                    AppendItem(acc, t.Value);
                }
                break;
            }
        }
    }

    private static byte[] EncodeKey(CborValue k)
    {
        if (!k.IsUInt)
            throw new AstboxError(E.InvalidCbor, "map keys must be integers");
        return EncodeUint(k.AsUInt);
    }

    /// <summary>Canonically encode an ASTBOX metadata object.</summary>
    public static byte[] Dumps(CborValue obj)
    {
        var acc = new List<byte>(256);
        AppendItem(acc, obj);
        return acc.ToArray();
    }
}

File: rust/crates/astbox-core/src/cbor_det.rs
// Copyright 2026 Astenyx-Git
// SPDX-License-Identifier: MPL-2.0 OR AGPL-3.0-only
//! Deterministic (canonical) CBOR encode/decode for ASTBOX v1 metadata.
//! Port of Astbox.Core/CborDet.cs (ASTBOX-v1.0-03 §13-19):
//! RFC 8949 deterministic encoding restricted to unsigned ints, byte strings,
//! UTF-8 text, arrays and maps; floats/tags/indefinite lengths/non-minimal
//! integers/duplicate or out-of-canonical-order map keys are rejected.

use crate::errors::{AstboxError, E};
use crate::Result;

/// Decoded CBOR value model (UInt | Bytes | Text | Array | Map).
/// Map entries are kept in decoded (canonical) order.
#[derive(Debug, Clone, PartialEq)]
pub enum CborValue {
    UInt(u64),
    Bytes(Vec<u8>),
    Text(String),
    Array(Vec<CborValue>),
    Map(Vec<(CborValue, CborValue)>),
}

impl CborValue {
    pub fn uint(v: u64) -> CborValue {
        CborValue::UInt(v)
    }
    pub fn bytes(b: impl Into<Vec<u8>>) -> CborValue {
        CborValue::Bytes(b.into())
    }
    pub fn text(s: impl Into<String>) -> CborValue {
        CborValue::Text(s.into())
    }
    pub fn arr(items: Vec<CborValue>) -> CborValue {
        CborValue::Array(items)
    }
    /// Map from (uint key, value) pairs, port of CborValue.Map(params (ulong, CborValue)[]).
    pub fn map_u(pairs: &[(u64, CborValue)]) -> CborValue {
        CborValue::Map(
            pairs
                .iter()
                .map(|(k, v)| (CborValue::UInt(*k), v.clone()))
                .collect(),
        )
    }

    pub fn is_uint(&self) -> bool {
        matches!(self, CborValue::UInt(_))
    }
    pub fn is_bytes(&self) -> bool {
        matches!(self, CborValue::Bytes(_))
    }
    pub fn is_text(&self) -> bool {
        matches!(self, CborValue::Text(_))
    }
    pub fn is_array(&self) -> bool {
        matches!(self, CborValue::Array(_))
    }
    pub fn is_map(&self) -> bool {
        matches!(self, CborValue::Map(_))
    }

    pub fn as_uint(&self) -> u64 {
        match self {
            CborValue::UInt(v) => *v,
            _ => 0,
        }
    }
    pub fn as_bytes(&self) -> &[u8] {
        match self {
            CborValue::Bytes(b) => b,
            _ => &[],
        }
    }
    pub fn as_text(&self) -> &str {
        match self {
            CborValue::Text(s) => s,
            _ => "",
        }
    }
    pub fn items(&self) -> &[CborValue] {
        match self {
            CborValue::Array(v) => v,
            _ => &[],
        }
    }
    pub fn entries(&self) -> &[(CborValue, CborValue)] {
        match self {
            CborValue::Map(m) => m,
            _ => &[],
        }
    }
}

pub struct CborDet;

const MAX_DEPTH: usize = 64;

struct Reader<'a> {
    data: &'a [u8],
    pos: usize,
}

impl<'a> Reader<'a> {
    fn remaining(&self) -> usize {
        self.data.len() - self.pos
    }

    fn take(&mut self, n: usize) -> Result<&'a [u8]> {
        if self.pos + n > self.data.len() {
            return Err(AstboxError::new(E::InvalidCbor, "truncated CBOR item"));
        }
        let s = &self.data[self.pos..self.pos + n];
        self.pos += n;
        Ok(s)
    }
}

/// (major, additionalInfo, length)
fn read_head(r: &mut Reader) -> Result<(u8, u8, u64)> {
    if r.remaining() < 1 {
        return Err(AstboxError::new(E::InvalidCbor, "truncated CBOR item"));
    }
    let b0 = r.data[r.pos];
    r.pos += 1;
    let major = b0 >> 5;
    let ai = b0 & 0x1F;
    let length: u64 = match ai {
        0..=23 => ai as u64,
        24 => r.take(1)?[0] as u64,
        25 => {
            let b = r.take(2)?;
            u16::from_be_bytes([b[0], b[1]]) as u64
        }
        26 => {
            let b = r.take(4)?;
            u32::from_be_bytes([b[0], b[1], b[2], b[3]]) as u64
        }
        27 => {
            let b = r.take(8)?;
            u64::from_be_bytes([
                b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7],
            ])
        }
        _ => {
            return Err(AstboxError::new(
                E::InvalidCbor,
                "indefinite-length item forbidden",
            ))
        }
    };
    Ok((major, ai, length))
}

fn check_minimal(ai: u8, length: u64) -> Result<()> {
    // Reject non-minimal integer encodings (RFC 8949 4.2.1).
    if ai == 24 && length < 24 {
        return Err(AstboxError::new(E::NonCanonicalCbor, "non-minimal uint encoding"));
    }
    if ai == 25 && length < 0x100 {
        return Err(AstboxError::new(E::NonCanonicalCbor, "non-minimal uint encoding"));
    }
    if ai == 26 && length < 0x1_0000 {
        return Err(AstboxError::new(E::NonCanonicalCbor, "non-minimal uint encoding"));
    }
    if ai == 27 && length < 0x1_0000_0000 {
        return Err(AstboxError::new(E::NonCanonicalCbor, "non-minimal uint encoding"));
    }
    Ok(())
}

fn decode_item(r: &mut Reader, depth: usize) -> Result<CborValue> {
    if depth > MAX_DEPTH {
        return Err(AstboxError::new(E::InvalidCbor, "CBOR nesting too deep"));
    }
    if r.remaining() < 1 {
        return Err(AstboxError::new(E::InvalidCbor, "truncated CBOR item"));
    }
    let (major, ai, length) = read_head(r)?;

    if major == 0 {
        check_minimal(ai, length)?;
        return Ok(CborValue::UInt(length));
    }

    match major {
        1 => Err(AstboxError::new(
            E::InvalidCbor,
            "negative CBOR integer forbidden",
        )),

        2 => {
            if length > usize::MAX as u64 {
                return Err(AstboxError::new(E::InvalidCbor, "byte string too large"));
            }
            Ok(CborValue::Bytes(r.take(length as usize)?.to_vec()))
        }

        3 => {
            if length > usize::MAX as u64 {
                return Err(AstboxError::new(E::InvalidCbor, "text string too large"));
            }
            let raw = r.take(length as usize)?;
            let text = std::str::from_utf8(raw).map_err(|_| {
                AstboxError::new(E::InvalidCbor, "text string is not UTF-8")
            })?;
            Ok(CborValue::Text(text.to_string()))
        }

        4 => {
            if length > usize::MAX as u64 {
                return Err(AstboxError::new(E::InvalidCbor, "array too large"));
            }
            let n = length as usize;
            let mut arr = Vec::with_capacity(n.min(1024));
            for _ in 0..n {
                arr.push(decode_item(r, depth + 1)?);
            }
            Ok(CborValue::Array(arr))
        }

        5 => {
            if length > usize::MAX as u64 {
                return Err(AstboxError::new(E::InvalidCbor, "map too large"));
            }
            let n = length as usize;
            let mut result: Vec<(CborValue, CborValue)> = Vec::with_capacity(n.min(1024));
            let mut prev_encoded: Option<(usize, Vec<u8>)> = None;
            for _ in 0..n {
                let key_start = r.pos;
                let key = decode_item(r, depth + 1)?;
                let key_end = r.pos;
                let value = decode_item(r, depth + 1)?;
                if result.iter().any(|(k, _)| *k == key) {
                    return Err(AstboxError::new(
                        E::DuplicateCborKey,
                        format!("duplicate map key {}", cbor_debug(&key)),
                    ));
                }
                result.push((key, value));
                // canonical key order: keys sorted by (encoded length, bytes)
                let key_len = key_end - key_start;
                let key_bytes = r.data[key_start..key_end].to_vec();
                if let Some((prev_len, prev_bytes)) = &prev_encoded {
                    if key_len < *prev_len
                        || (key_len == *prev_len && key_bytes.as_slice() <= prev_bytes.as_slice())
                    {
                        return Err(AstboxError::new(
                            E::NonCanonicalCbor,
                            "map keys not in canonical order",
                        ));
                    }
                }
                prev_encoded = Some((key_len, key_bytes));
            }
            Ok(CborValue::Map(result))
        }

        // majors 6 (tags) and 7 (floats/specials) are forbidden
        _ => Err(AstboxError::new(
            E::InvalidCbor,
            format!("CBOR major type {} forbidden in ASTBOX metadata", major),
        )),
    }
}

/// Debug formatting mirroring CborValue.ToString() for error messages.
fn cbor_debug(v: &CborValue) -> String {
    match v {
        CborValue::UInt(u) => format!("uint({})", u),
        CborValue::Bytes(b) => format!("bytes[{}]", b.len()),
        CborValue::Text(s) => format!("text({})", s.chars().count()),
        CborValue::Array(a) => format!("array[{}]", a.len()),
        CborValue::Map(m) => format!("map[{}]", m.len()),
    }
}

fn encode_uint(value: u64) -> Vec<u8> {
    if value < 24 {
        vec![value as u8]
    } else if value < 0x100 {
        vec![0x18, value as u8]
    } else if value < 0x1_0000 {
        let mut b = vec![0x19];
        b.extend_from_slice(&(value as u16).to_be_bytes());
        b
    } else if value < 0x1_0000_0000 {
        let mut b = vec![0x1A];
        b.extend_from_slice(&(value as u32).to_be_bytes());
        b
    } else {
        let mut b = vec![0x1B];
        b.extend_from_slice(&value.to_be_bytes());
        b
    }
}

fn encode_head(major: u8, length: u64) -> Vec<u8> {
    let m = major << 5;
    if length < 24 {
        vec![(m | length as u8)]
    } else if length < 0x100 {
        vec![m | 24, length as u8]
    } else if length < 0x1_0000 {
        let mut b = vec![m | 25];
        b.extend_from_slice(&(length as u16).to_be_bytes());
        b
    } else if length < 0x1_0000_0000 {
        let mut b = vec![m | 26];
        b.extend_from_slice(&(length as u32).to_be_bytes());
        b
    } else {
        let mut b = vec![m | 27];
        b.extend_from_slice(&length.to_be_bytes());
        b
    }
}

fn encode_key(k: &CborValue) -> Result<Vec<u8>> {
    match k {
        CborValue::UInt(v) => Ok(encode_uint(*v)),
        _ => Err(AstboxError::new(E::InvalidCbor, "map keys must be integers")),
    }
}

fn append_item(acc: &mut Vec<u8>, obj: &CborValue) -> Result<()> {
    match obj {
        CborValue::UInt(v) => {
            acc.extend_from_slice(&encode_uint(*v));
        }

        CborValue::Bytes(b) => {
            acc.extend_from_slice(&encode_head(2, b.len() as u64));
            acc.extend_from_slice(b);
        }

        CborValue::Text(s) => {
            // metadata strings are NFC-normalized
            use unicode_normalization::UnicodeNormalization;
            let norm: String = s.nfc().collect();
            let data = norm.as_bytes();
            acc.extend_from_slice(&encode_head(3, data.len() as u64));
            acc.extend_from_slice(data);
        }

        CborValue::Array(items) => {
            acc.extend_from_slice(&encode_head(4, items.len() as u64));
            for item in items {
                append_item(acc, item)?;
            }
        }

        CborValue::Map(entries) => {
            let mut ordered: Vec<(&CborValue, &CborValue, Vec<u8>)> = entries
                .iter()
                .map(|(k, v)| Ok((k, v, encode_key(k)?)))
                .collect::<Result<Vec<_>>>()?;
            ordered.sort_by(|a, b| a.2.cmp(&b.2));
            acc.extend_from_slice(&encode_head(5, ordered.len() as u64));
            for (_, value, enc_key) in ordered {
                acc.extend_from_slice(&enc_key);
                append_item(acc, value)?;
            }
        }
    }
    Ok(())
}

impl CborDet {
    /// Strictly decode canonical CBOR; trailing bytes forbidden.
    pub fn loads(data: &[u8]) -> Result<CborValue> {
        let mut r = Reader { data, pos: 0 };
        let obj = decode_item(&mut r, 0)?;
        if r.remaining() != 0 {
            return Err(AstboxError::new(
                E::InvalidCbor,
                "trailing bytes after CBOR item",
            ));
        }
        Ok(obj)
    }

    /// Canonically encode an ASTBOX metadata object.
    pub fn dumps(obj: &CborValue) -> Result<Vec<u8>> {
        let mut acc = Vec::with_capacity(256);
        append_item(&mut acc, obj)?;
        Ok(acc)
    }
}
